### PR TITLE
[RNMobile] Refactor: gallery to nested image blocks

### PR DESCRIPTION
**Pull request by dratwas**
## Description
Refactors the Gallery block to be a wrapper around Image blocks using InnerBlocks. This is a mobile implementation of https://github.com/WordPress/gutenberg/pull/25940

### What is done:
In order to make it works, I had to do a lot of changes. The main change is related to the "Tile" component. Till now the Gallery block used the Tile component to set the correct layout of the gallery, however, now we use inner blocks for rendering and there is no possibility to use it in the same way since the InnerBlock component is responsible for the layout. I moved the code from gallery/tile to the block-list/grid-item and use it there. In order to make it works, I also had to add the `gridProperties` prop to set the correct number of columns (there could be more properties in the future).

In the base PR the author moved some code from media-placeholder to the gallery/edit and I had also done the same to be compatible with that.

Another important thing is that the image block is not prepared to be used in the galley. I mean that images in the gallery have the same height etc so I had also make some changes in the image block. I added a prop `isGallery` and` imageCrop` to determine how the image should look like.

I couldn't use the `useParentAttributes` to get inherited attributes from the gallery because the image/edit.native.js is a class. I implemented it in the class way (using the componentDidUpdate) however, it will be probably changed by the web team and should be implemented here as well - I describe it below.

I also had to add some checks in the `link-settings/index.native.js` to prevent setting the URL when it's not necessary because it was setting the wrong URL when using the nested URL from the gallery. 

I implemented the possibility to use the `blockType.usesContext` as it is done on the web. It was used to pass down some attributes from the gallery to the image.

The thing that is still under development and I believe we should wait for the final implementation is how the inner blocks should behave when we change some attributes in the gallery. There are 3 attributes that are in the gallery and also in the image block:
- inkDestination
- linkTarget
- sizeSlug

The discussion about that https://github.com/WordPress/gutenberg/pull/25940#discussion_r574300020

At this moment they are passed down through the context (that's why I had to add the possibility to use the BlockContext), however, there is still discussion on how it should be done and how the inner blocks should act when we change one of the listed attributes. 

I really suggest waiting for the final decision since I had to change the implementation a few times in the meantime because of the new review.

### What should be done:

First of all, as I mentioned we should wait till the web team decides about all implementation details because I don't think that it makes sense to implement it parallelly since they change his approach quite frequently.

Then this PR should be rebased to the latest `refactor/gallery-to-nested-image-blocks`

Then depending on the web team's decision about how to handle the inherited attributes that are in the gallery block and the image block as well, made changes in the image/edit.native.js.

The important thing is also that the web team wants to keep the old implementation and new implementation together:
https://github.com/WordPress/gutenberg/pull/25940#issuecomment-777256259

## How has this been tested?
All sanity tests for gallery 
Gallery-1

- [ ] Gallery block - Close/Re-open post with an ongoing image upload - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc001)
- [ ] Gallery block - Close post with an ongoing image upload - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc002)

Gallery-2

- [ ] Gallery block - Insert image from device (failing) - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc006)
- [ ] Gallery block - Insert image from device (cancel) - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc009)

Gallery-3

- [ ] Gallery block - Add Gallery Caption - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc003)
- [ ] Gallery block - Add Gallery Image Caption - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc004)
- [ ] Gallery block - Try adding same images from WP Media library and moving the images around - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc010)

Gallery-4

- [ ] Gallery block - Settings: Link to - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc012)
- [ ] Gallery block - Settings: Column number - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc013)
- [ ] Gallery block - Settings: Crop images - [steps](https://github.com/wordpress-mobile/test-cases/blob/master/test-cases/gutenberg/gallery.md#tc014)

## Screenshots 

ezgif com-gif-maker

## Types of changes
New feature (non-breaking change which adds functionality)
Refactors the Gallery block to be a wrapper around Image blocks using InnerBlocks. 

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Review by hc0503 — APPROVED** [review]

**Comment by dratwas** [comment]
@iamthomasbishop 
Answering the comment from base PR https://github.com/WordPress/gutenberg/pull/25940#issuecomment-725027248

Moved that discussion here to have it in the right place since it is related only to the mobile part.

>@dratwas Assuming you are referring to the dashed child borders, why do we need to remove them? Maybe I'm missing context, but I would expect the bordering to look/work the same way as other innerBlocks do (the screenshot you shared in your comment looks correct).

Maybe I should as in a bit different way, sorry for that. My question should be: "Do we want to fit the innerBlocks with image blocks inside to keep the previous gallery look, or do we want to have default innerBlocks look"

Previous Gallery | Default InnerBlocks
--- | ---
Simulator Screen Shot - iPhone 11 Pro - 2020-11-13 at 12 14 25 | Simulator Screen Shot - iPhone 11 Pro - 2020-11-13 at 12 08 06

I created the test build for you here: https://github.com/wordpress-mobile/WordPress-iOS/pull/15284#issuecomment-726367909

**Comment by iamthomasbishop** [comment]
> Maybe I should as in a bit different way, sorry for that. My question should be: "Do we want to fit the innerBlocks with image blocks inside to keep the previous gallery look, or do we want to have default innerBlocks look"

@dratwas Ahh okay, makes sense! My preference would be to utilize the standard innerBlocks UI (with the dashed child borders), so the second screenshot on your [last comment](https://github.com/WordPress/gutenberg/pull/26823#issuecomment-726709097) ("default innerBlocks") looks correct. 

Side note: I was trying to test the build linked in wordpress-mobile/WordPress-iOS#15284 and I'm not seeing the innerBlocks style — maybe it's behind a flag?

**Comment by dratwas** [comment]
Thanks for the review @iamthomasbishop. I created a new build with the latest changes - https://github.com/wordpress-mobile/WordPress-iOS/pull/15284#issuecomment-726367909

Please test it and leave your thoughts here :)

Side note: I feel like after change from the old implementation (Array.map) to the innerBlocks (FlatList) the performance is noticeably worse. I can observe that when changing the count of images in a row when there are more than 5 images in the gallery.

**Comment by dratwas** [comment]
Hey, @iamthomasbishop I changed the inline cells with `Apply to all images` and `Cancel` to the ActionSheet. Could you please take a look at the newest build? :) https://github.com/wordpress-mobile/WordPress-iOS/pull/15284#issuecomment-726367909

**Comment by iamthomasbishop** [comment]
@dratwas That definitely feels better! Which component are we using for this on Android? I'm guessing BottomSheet?

One request: can we change the title/header label ("Gallery block options") to a more specific message/alert text like "Apply to all images in this gallery?". Here's an example (please forgive the quick dirty mock-up that I made on my phone 😅):

image

One thing that seems like a bug: I'm also seeing an odd looping issue with when I first tap on the block settings icon (but if I kill the app and resume, the issue goes away). Here's a video: https://cloudup.com/cHML3GFceui . Note: it's hard to tell but after I opened the settings sheet, the alert/actionSheet popped up and continued reopening immediately upon closing it.

**Comment by dratwas** [comment]
Hey, @iamthomasbishop I created new builds for iOS and Android :) Could you take a look?

I fixed the loop and changed the title of the picker.

https://github.com/wordpress-mobile/WordPress-Android/pull/13423#issuecomment-730258624
https://github.com/wordpress-mobile/WordPress-iOS/pull/15284#issuecomment-726367909

**Comment by iamthomasbishop** [comment]
@dratwas Much much better, thank you! I have one question and it's a minor thing, not a big blocker. I'm curious if we can "pause" changes to the bottomSheet (setting label and sheet height) until _after_ the selection is made on the actionSheet? 

For example, let's say I have the `Link to` setting set to `None`. If I tap on the `Link to` cell and select another option (like `Attachment Page`) from the initial actionSheet, the new option is applied immediately (visible on the sheet) and the height of the sheet changes — _before_ the confirmation step. 

If possible, I'd like to "pause" applying the change until the user has tapped "Apply to all images" on the confirmation sheet. Alternatively, if they tap "Cancel", there should be no visible change.

**Comment by dratwas** [comment]
Hey @iamthomasbishop 
> I'm curious if we can "pause" changes to the bottomSheet (setting label and sheet height) until after the selection is made on the actionSheet?

Unfortunately in the current implementation - NO :(

This is also a shared code between the web and mobile. It sets attributes immediately and then if the user clicks "cancel" it reverts them. When the user clicks "Apply" then those changes are applied to all images inside. 

If it's not a blocker I would go with the current implementation and create an issue for that. This PR is only a mobile part of the gallery refactoring and I'm afraid that it's already too big. What do you think?

File: packages/block-editor/src/components/media-placeholder/index.native.js [hunk 1]
```diff
@@ -163,7 +144,9 @@ function MediaPlaceholder( props ) {
 							accessibilityRole={ 'button' }
 							accessibilityHint={ accessibilityHint }
 							onPress={ ( event ) => {
-								props.onFocus( event );
+								if ( props.onFocus ) {
```
Comment (lukewalczak) [inline packages/block-editor/src/components/media-placeholder/index.native.js:147]: Maybe add `onFocus` to props destructuring above 🤔

File: packages/block-library/src/gallery/dirty-image-options.android.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+const options = [ applyToAllImagesOption ];
+
+export default ( { applyImageOptions, cancelImageOptions, isVisible } ) => {
+	const pickerRef = useRef();
+	const [ apply, setApply ] = useState(false);
+
+	useEffect( () => {
+		InteractionManager.runAfterInteractions( () => {
+			if ( isVisible && pickerRef.current ) {
```
Comment (lukewalczak) [inline packages/block-library/src/gallery/dirty-image-options.android.js:26]:
```suggestion
			if ( isVisible && pickerRef?.current ) {
```

File: packages/block-library/src/gallery/dirty-image-options.ios.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+		<Picker
+			ref={ pickerRef }
+			options={ options }
+			onChange={ onChange }
+			onCancel={ cancelImageOptions }
+			leftAlign={ true }
+			title={ __( 'Apply to all images in this gallery?' ) }
+		/>
+	);
+};
```
Comment (lukewalczak) [inline packages/block-library/src/gallery/dirty-image-options.ios.js:45]: `dirty-image-options.ios.js` and `dirty-image-options.android.js` look alike, can we somehow unify both files?

File: packages/block-library/src/gallery/gallery-styles.native.scss [hunk 4]
```diff
@@ -2,6 +2,10 @@
 	margin-bottom: 16px;
 }
 
+.galleryAppender {
+	padding-top: 16;
```
Comment (lukewalczak) [inline packages/block-library/src/gallery/gallery-styles.native.scss:6]: Use grid variable

**Review by lukewalczak — CHANGES_REQUESTED** [review]

**Comment by iamthomasbishop** [comment]
> If it's not a blocker I would go with the current implementation and create an issue for that.

That's totally fine, thanks @dratwas!

**Comment by chipsnyder** [comment]
Closing this PR in favor of the updated mobile PR https://github.com/WordPress/gutenberg/pull/31306

**Pull request closed without merging**